### ISSUE TITLE
1.1.13: ext/atomic_reference.c: No CAS operation available for this platform

### ISSUE THREAD

**Issue by marksolaris** [headius/ruby-atomic#30]
After fixing the -march=native issue the install bombed on ext/atomic_reference.c

```
# gem install ${GEM_PATH}/atomic-1.1.13.fixed.gem
Building native extensions.  This could take a while...
ERROR:  Error installing /usr/local/src/ruby/atomic-1.1.13.fixed.gem:
        ERROR: Failed to build gem native extension.

    /usr/local/bin/ruby extconf.rb
creating Makefile

make "DESTDIR="
compiling atomic_reference.c
"atomic_reference.c", line 58: #error: No CAS operation available for this platform
cc: acomp failed for atomic_reference.c
*** Error code 2
The following command caused the error:
cc -I. -I/usr/local/include/ruby-2.0.0/i386-solaris2.11 -I/usr/local/include/ruby-2.0.0/ruby/backward -I/usr/local/include/ruby-2.0.0 -I. -D_FILE_OFFSET
_BITS=64  -KPIC -I/usr/local/include  -L/usr/local/lib -DHAVE_UCRED_H  -KPIC   -o atomic_reference.o -c atomic_reference.c
make: Fatal error: Command failed for target `atomic_reference.o'
```

due to this section of the file:

```
48  static VALUE ir_compare_and_set(volatile VALUE self, VALUE expect_value, VALUE new_value) {
49  #if __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ >= 1050
50      if (OSAtomicCompareAndSwap64(expect_value, new_value, &DATA_PTR(self))) {
51      return Qtrue;
52      }
53  #elif HAVE_GCC_CAS
54      if (__sync_bool_compare_and_swap(&DATA_PTR(self), expect_value, new_value)) {
55      return Qtrue;
56      }
57  #else
58  #error No CAS operation available for this platform
59  #endif
60      return Qfalse;
61  }
```

Issue #23 walked up to the problem and didn't encounter it as they were using gcc.

**Comment by marksolaris** [comment]
So the choices in the code are; use OSX, or use GCC. To be more portable the atomic code should be agnostic about what it runs on. Some discussions are here:

http://stackoverflow.com/questions/1158374/portable-compare-and-swap-atomic-operations-c-c-library

and the Open Portable Atomics library is by definition pretty open and portable

https://trac.mcs.anl.gov/projects/openpa/wiki/Overview

this demo code shows how: https://trac.mcs.anl.gov/projects/openpa/browser/test/sanity.c

**Comment by marksolaris** [comment]
This "appears" to have passed compilation... but it's not tested. How to test? I'm guessing that it's being passed 64bit ints as function arguments. The URL inline has a list of other functions for other data types.

```
atomic-1.1.13.gem.cas.diff
--- ext/atomic_reference.orig.c Wed Aug 21 18:53:27 2013
+++ ext/atomic_reference.c      Wed Aug 21 19:08:30 2013
@@ -11,6 +11,9 @@
 // limitations under the License.

 #include <ruby.h>
+#if defined(__sun)
+#include <atomic.h>
+#endif

 static void ir_mark(void *value) {
     rb_gc_mark_maybe((VALUE) value);
@@ -54,6 +57,12 @@
     if (__sync_bool_compare_and_swap(&DATA_PTR(self), expect_value, new_value)) {
        return Qtrue;
     }
+#elif defined(__sun)
+    /* assuming these are 64bit ints */
+    /* Fun with data types http://docs.oracle.com/cd/E19253-01/816-5168/6mbb3hr1v/index.html */
+    if (atomic_cas_64(&DATA_PTR(self), expect_value, new_value)) {
+        return Qtrue;
+    }
 #else
 #error No CAS operation available for this platform
 #endif
```

This is how it's applied to a default atomic-1.1.13.gem file. The checksums.yaml.gz file wasn't repacked.

```
cd /var/tmp
rm -rf /var/tmp/atomic_fix
mkdir /var/tmp/atomic_fix
cd /var/tmp/atomic_fix
tar -xf ${GEM_PATH}/atomic-1.1.13.gem
gunzip -c data.tar.gz | tar -tf - > tarfiles
gunzip -c data.tar.gz | tar -xf -
perl -pe 's#-march=native##' -i ext/extconf.rb
patch -p0 < /usr/local/src/ruby/atomic-1.1.13.gem.cas.diff
tar -cf - -I tarfiles | gzip -9c > data.tar.gz
tar -cf ${GEM_PATH}/atomic-1.1.13.fixed.gem data.tar.gz metadata.gz
gem install ${GEM_PATH}/atomic-1.1.13.fixed.gem
```

**Comment by orderthruchaos** [comment]
I'd like to recommend a slight change to the code.  The following assumes that VALUE is defined as uintptr_t:

``` cc
+/* Assuming VALUE is uintptr_t */
+#elif defined(__sun)
+/* Based on the definition of uintptr_t from /usr/include/sys/int_types.h */
+#if defined(_LP64) || defined(_I32LPx)
+    /* 64-bit: uintptr_t === unsigned long */
+    if (atomic_cas_ulong(&DATA_PTR(self), expect_value, new_value)) {
+        return Qtrue;
+    }
+#else
+    /* 32-bit: uintptr_t === unsigned int */
+    if (atomic_cas_uint(&DATA_PTR(self), expect_value, new_value)) {
+        return Qtrue;
+    }
+#endif
```

This code should work for both SPARC and Intel-compatible architectures.

**Comment by marksolaris** [comment]
Sounds good, it wasn't actually clear what VALUE was being given at, my patch was simply gambling :)

**Comment by headius** [comment]
Should be fixed by #31, which is in 1.1.14 now.
